### PR TITLE
Bigtable: 21. Refactor Batching - Move retries behind batching

### PR DESCRIPTION
**Pull request by igorbernstein2**
This is the first step to enable batch flushing. Currently, the outermost callable for bulk mutations is a RetryingCallable. Although this composes nicely with automatic batching, it makes it very difficult to reason about manual flushing of batches. For example: what happens when multiple entries fail after a flush? should batcher retry each entry individually? should it spool them? etc. By moving the retries after batching, flushing becomes easier: on flush, the batcher sends all of its batches to the retry logic and waits for all batches to either succeed or run out of attempts. 

The implementation uses the retrying infrastructure directly, bypassing Callable.retrying & RetryingCallable. This is necessary because I need to track state of completed entries. This state is maintained in the new MutateRowsAttemptCallable class, which tracks incomplete mutation entries.  The class also transforms the results of the RPC to be a bit more ergonomic:
- upon either RPC or entry failures a MutateRowsException is thrown. This exception can be inspected to identify failed mutations.
- Since a successful MutateRowsResponse holds no interesting information, it is replaced with a Void.

At some point in the future I would like to try to generalize `MutateRowsAttemptCallable` to simply wrap something like a `BatchingDescriptor`. This would allow the GAPIC generated batching configuration to be used for implementing partial retries on batches.

The PR is broken up into 3 commits:
1. Adds the new attempt `Callable`, which is equivalent of gax's AttemptCallable. It contains all of the logic for composing a request.
2. Wraps the new `Callable` in a `UnaryCallable` which uses gax's `RetryingExecutor` to manage retry attempts
3. Update the Bigtable chain to use the new functionality

It would be easiest to review this a commit at a time.

### CODE REVIEW

**Comment by igorbernstein2** [comment]
Rebased. This is ready for review. @garrettjonesgoogle & @vam-google can you take a look when you have a moment?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    ApiException entryError = createMissingEntryError(throwable);
+    List<FailedMutation> allFailures;
+
+    if (entryError.isRetryable()) {
+      allFailures =
+          Lists.newArrayListWithCapacity(
+              permanentFailures.size() + currentRequest.getEntriesCount());
+      allFailures.addAll(permanentFailures);
+    } else {
+      allFailures = permanentFailures;
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java]: I refactored the logic, let me know if it makes more sense. I was being a little too clever.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    if (!allFailures.isEmpty()) {
+      boolean isRetryable = builder.getEntriesCount() > 0;
+      throw new MutateRowsException(null, allFailures, isRetryable);
+    }
+  }
+
+  /**
+   * Remap the entry index in the current attempt back to its original index in the first request.
+   */
+  private int remapIndex(int index) {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java]: done

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    return (indexMap != null) ? indexMap.get(index) : index;
+  }
+
+  /** Convert an entry's status from a protobuf to an {@link ApiException}. */
+  private ApiException createEntryError(com.google.rpc.Status protoStatus) {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java]: no, it needs access to retryableCodes. But the next method can be

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        gaxStatusCode,
+        retryableCodes.contains(gaxStatusCode.getCode()));
+  }
+
+  /**
+   * Create a synthetic {@link ApiException} for an individual entry. When the entire RPC fails, it
+   * implies that all entries failed as well. This helper is used to make that behavior explicit.
+   * The generated exception will have the overall error as its cause.
+   */
+  private ApiException createMissingEntryError(Throwable overallRequestError) {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java]: done

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptor.java [hunk 5]
```diff
@@ -76,44 +52,49 @@ public PartitionKey getBatchPartitionKey(MutateRowsRequest request) {
...
-    for (BatchedRequestIssuer<MutateRowsResponse> responder : batch) {
-      responder.setException(throwable);
+    int i = 0;
+    for (BatchedRequestIssuer<Void> issuer : batch) {
+      for (int j = 0; j < issuer.getMessageCount(); j++) {
+        FailedMutation failure = errorsByIndex.get(i++);
+        if (failure == null) {
+          issuer.setResponse(null);
+        } else {
+          issuer.setException(failure.getError());
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptor.java]: I updated the code to be more explicit that we want the last error. In practice, this doesn't matter. Each issuer will have exactly 1 message.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/retrying/ApiResultRetryAlgorithm.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import org.threeten.bp.Duration;
+
+public class ApiResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<ResponseT> {
+  // Duration to sleep on if the error is DEADLINE_EXCEEDED.
+  public static final Duration DEADLINE_SLEEP_DURATION = Duration.ofMillis(1);
+
+  @Override
+  public TimedAttemptSettings createNextAttempt(
+      Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings prevSettings) {
+    if (prevThrowable != null && prevThrowable instanceof DeadlineExceededException) {
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/retrying/ApiResultRetryAlgorithm.java]: It's a copy of https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/ApiResultRetryAlgorithm.java. I'd prefer to keep them identical

**Comment by igorbernstein2** [comment]
rebased

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallableTest.java [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                    .setStatus(PERMENANT_ERROR_STATUS_PROTO))
+            .build());
+
+    MutateRowsAttemptCallable attemptCallable =
+        new MutateRowsAttemptCallable(innerCallable, request, callContext, retryCodes);
+    attemptCallable.setExternalFuture(parentFuture);
+
+    // Make the first call
+    attemptCallable.call();
+
```
Comment (kevinsi4508) [inline google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallableTest.java]: Verifying which entry get retried?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptor.java [hunk 8]
```diff
@@ -76,44 +52,55 @@ public PartitionKey getBatchPartitionKey(MutateRowsRequest request) {
...
+        FailedMutation failure = errorsByIndex.get(i++);
+        if (failure != null) {
+          lastError = failure.getError();
+        }
+      }
+
+      if (lastError == null) {
+        issuer.setResponse(null);
+      } else {
+        issuer.setException(lastError);
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptor.java]: I added a comment to clarify.

File: google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallableTest.java (hunk 7 above)
Comment (igorbernstein2) [reply google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallableTest.java]: done

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+      } else {
+        // Schedule the mutation entry for the next RPC by adding it to the request builder and
+        // recording it's original index
+        newOriginalIndexes.add(origIndex);
+        builder.addEntries(lastRequest.getEntries(i));
+      }
+    }
+
+    currentRequest = builder.build();
+    originalIndexes = newOriginalIndexes;
```
[the remark this answers is not in the document]
Comment (igorbernstein2) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsAttemptCallable.java]: Can you clarify what you mean by tracking individual requests & results? Do you mean keeping a map of each entry and its status and then iterating through that list for every retry checking if it eligible for a retry? Or something else? Also can you clarify how it helps debugging?

I'm not sure what you mean about the correctness of 3rd call. At the beginning of each call the following invariants should ensure that every attempt is correct
- originalIndexes should have a 1:1 correspondence with currentRequest.entries
- currentRequest.entries + permanentFailures = original request
- last MutateRowsException.failedMutations - permanentFailures = currentRequest

**Comment by igorbernstein2** [comment]
Thanks for reviewing, please merge when ready

**Pull request merged**